### DISCUSSION TITLE
Handling errors globally

### DISCUSSION THREAD

**Discussion by lukeredpath** [pointfreeco/swift-composable-architecture#383] [Ideas]
I think there has been some discussion about this on the Swift Forums before but I have been thinking about ways to handle certain types of error in a global way, when those errors could come from effects deep within the domain hierarchy.

On a fundamental level, it feels like if a child domain needs to communicate in some way with an ancestor domain, this should be treated as a side effect - it doesn't know about the parent or ancestor domains so all it can really do is broadcast something for other listeners to observe. 

To take a concrete example, imagine an API client that makes authenticated API requests - if for some reason the auth token has expired the API will return a 401 response and we should ask the user to log in again. 

We have a feature domain in our app that is responsible for managing session state and presenting the login screen when necessary but other feature domains deeper within our state hierarchy do not know about this. 

I can think of a few ways to handle this, such as having a reducer at the session feature domain level handling every possible API client response action in the app but this feels like a lot of boilerplate and not very ergonomic. 

So what if...the API client simply republished its errors as a Notification using NotificationCenter? The session domain could then start a long running effect that observes these notifications using a notification publisher and maps each notification to an action. The session reducer then only needs to handle a single action where it can check for a not authorised error and trigger a login.

Does this seem like a reasonable approach? Are there any downsides that I'm missing? Is there an alternative solution that would be better? 

I feel like this technique could be applied to a number of scenarios where you have some root level domain that needs to deal with events from deep within the hierarchy. Another example would be displaying toast banners - we can implement a toast system as a Composable feature domain using state and effects but we only want to have a single banner state in the root of our state hierarchy (because there's a single global banner queue for the whole app). Again features could publish events as a Notification and the banner domain could observe these in an effect of its own and trigger banners in response to these notifications.

**Comment by lukeredpath** [comment]
As I'm back at my desk I thought I'd update this with a few code examples.

👉🏻  [Update: see updated implementation that doesn't use NotificationCenter at all below](https://github.com/pointfreeco/swift-composable-architecture/discussions/383#discussioncomment-359851)

Firstly, we can write a general Publisher operator that intercepts and re-publishes errors as a `Notification` before passing it downstream:

```swift
extension Publisher {
    func republishErrors(
        as name: Notification.Name,
        userInfoKey: AnyHashable,
        to notificationCenter: NotificationCenter = .default
    ) -> AnyPublisher<Output, Failure> {
        return self.catch { error -> Fail<Output, Failure> in
            notificationCenter.post(
                name: name,
                object: nil,
                userInfo: [userInfoKey: error]
            )
            return Fail(outputType: Output.self, failure: error)
        }
        .eraseToAnyPublisher()
    }
}
```

Now we can write a more specific operator on `Effect` that re-publishes errors using a fixed notification name and key - these can be private as they are an implementation detail that we can keep encapsulated:

```swift
private let effectErrorUserInfoKey = "Effect.effectErrorUserInfoKey"

fileprivate extension Notification.Name {
    static let effectError: Self = .init("effectError")
}

extension Effect {
    func republishErrors() -> Self {
        republishErrors(
            as: .effectError,
            userInfoKey: effectErrorUserInfoKey,
            to: .default
        )
        .eraseToEffect()
    }
}
```

Finally we can define a custom publisher for these re-published errors as an extension on `NotificationCenter`:

```swift
extension NotificationCenter {
    func republishedEffectErrorsPublisher<E: Error>(of: E.Type) -> AnyPublisher<E, Never> {
        publisher(for: .effectError)
            .compactMap { $0.userInfo?[effectErrorUserInfoKey] as? E }
            .eraseToAnyPublisher()
    }
}
```

So now to put it to use - let's say we have some kind of HTTP client that can fail with an `HTTPClientError` - the implementation isn't important but the interface is a bit like this:

```swift
enum HTTPClientError {
  case notAuthorized
}

struct HTTPClient {
  var request: (HTTPRequest) -> Effect<HTTPResponse, HTTPClientError>
}
```

We can pass a client around in the environment and explicitly re-publish errors on a per-request basis. We can also do any custom error handling specific to this particular request that is relevant to our feature domain, but all HTTP errors will be re-published for other domains to intercept.

```swift
let reducer = Reducer<State, Action, Environment> { state, action, environment in 
  switch action {
    case .startSomeRequest:
      return environment
        .request(.someHTTPRequest)
        .republishErrors()
  }
}
```

We could even go one step further and introduce a "higher-order HTTP client" that always republishes its errors:

```swift
extension HTTPClient {
  func errorRepublishing() -> HTTPClient {
    .init { self.request($0).republishErrors() }
  }
}
```

Now the consumers of our `HTTPClient` don't even need to know about this republishing behaviour.

Finally, elsewhere in another feature domain, perhaps some kind of general error handling domain or maybe just our app domain, we can start a long running effect that listens for these errors, or perhaps filters out specific errors:

```swift
enum AppAction {
  case appFinishedLaunching
  case httpClientError(HTTPClientError)
}

struct AppState {
  var errorAlert: AlertState?
}

struct AppEnvironment {
  var httpClientErrors: Effect<HTTPClientError, Never>
}

extension AppEnvironment {
  static let live = AppEnvironment(
    httpClientErrors: NotificationCenter.default.republishedEffectErrorsPublisher(of: HTTPClientError.self)
  )
}

let appReducer = Reducer<AppState, AppAction, AppEnvironment> { state, action, environment in
  switch action {
    struct HTTPClientErrorPublisherID: Hashable {}

    case appFinishedLaunching:
      return environment
        .httpClientErrors
        .map(AppAction.httpClientError)
        .cancellable(id: HTTPClientErrorPublisherID())

    case let httpClientError(error):
      state.errorAlert = makeAlertForError(error) // for example
      return .none
  }
}

**Reply by lukeredpath** [reply]
Just to add - I realise there is still a subtly hidden coupling to `NotificationCenter` here because the `.republishErrors()` operator on `HTTPClient` uses `NotificationCenter` directly. This could be further abstracted away by having a more generic `republishTo()` function that isn't coupled directly to `NotificationCenter` - in fact the more I think about it the more I think you don't necessarily even need `NotificationCenter` at all and could possibly achieve this effect using a `Subject` 🤔

**Comment by lukeredpath** [comment]
Right, having given this a bit more thought about how to deal with the hidden `NotificationCenter` dependency in the above example, I've realised that the same kind of idea can be achieved more generally using a `PassthroughSubject`. 

You can pass the subject around in your `Environment` as an actual `PassthroughSubject` to parts of your domain that need to re-publish things or you can hide this implementation in a closure that takes an argument and forwards it to the subject inside a fire-and-forget effect.

Similarly you can pass the same subject - erased to a long-living effect - into domains that need to just subscribe to these re-published outputs or errors. This still allows two disconnected domains to communicate indirectly without knowing about the other.

Some updated code, firstly some `Effect` operators for re-publishing:

```swift
extension Effect {
    func republishErrors(to subject: PassthroughSubject<Failure, Never>) -> Effect<Output, Failure> {
        self.catch { error -> Effect<Output, Failure> in
            subject.send(error)
            return Effect(error: error)
        }
        .eraseToEffect()
    }

    func republishOutput(to subject: PassthroughSubject<Output, Never>) -> Effect<Output, Failure> {
        self.flatMap { value -> Effect<Output, Failure> in
            subject.send(value)
            return Effect(value: value)
        }
        .eraseToEffect()
    }
}
```

Our `AppEnvironment` now needs to hold both an effect and the subject (the former for subscribing to, the latter for publishing to):

```swift
struct AppEnvironment {
  var httpClient: HTTPClient
  var httpClientErrors: Effect<HTTPClientError, Never>
  var httpClientErrorSubject: PassthroughSubject<HTTPClientError, Never>
}
```

We can create this subject when we create our live environment (the mock version of this could allow us to pass in a `PassthroughSubject` explicitly from our test):

```swift
extension AppEnvironment {
  static var live: AppEnvironment {
    let httpClientErrorsSubject = PassthroughSubject<HTTPClientError, Never>()
    return AppEnvironment(
      httpClient: HTTPClient(),
      httpClientErrors: httpClientErrors.eraseToEffect(),
      httpClientErrorsSubject: httpClientErrorsSubject
    )
  }
}
```

Our feature reducer doesn't look much different to before, with the only difference being that we need to pass the `httpClientErrorsSubject` into each environment that needs it:

```swift
let reducer = Reducer<State, Action, Environment> { state, action, environment in 
  switch action {
    case .startSomeRequest:
      return environment.httpClient
        .request(.someHTTPRequest)
        .republishErrors(to: environment.httpClientErrorsSubject)
  }
}
```

Of course, we could improve the ergonomics here by returning to the high-level HTTP client idea from before and bundling this behaviour up in the `AppEnvironment`, so other domains do not need to know anything about the republishing or the subject:

```swift
extension HTTPClient {
  func errorRepublishing(to subject: PassthroughSubject<HTTPClientError, Never>) -> HTTPClient {
    .init { self.request($0).republishErrors(to: subject) }
  }
}

extension AppEnvironment {
  static var live: AppEnvironment {
    let httpClientErrorsSubject = PassthroughSubject<HTTPClientError, Never>()
    return AppEnvironment(
      httpClient: HTTPClient().errorRepublishing(to: httpClientErrorsSubject),
      httpClientErrors: httpClientErrors.eraseToEffect()
      // httpClientErrorsSubject: httpClientErrorsSubject - we wouldn't need this anymore
    )
  }
}
```

**Reply by nmccann** [reply]
Our team recently faced the same issue (needing a global error handling, primarily for booting the user back to login on certain HTTP statuses) and we independently ended up with a solution similar to your NotificationCenter-based approach. I was pretty sure we could do something similar using subjects, but wasn't familiar enough with Combine to manage it within the timeline we had. So thank you for sharing your `PassthroughSubject` based approach - that feels far less "hacky" than the `NotificationCenter` one.

**Reply by lukeredpath** [reply]
Glad it can be of help and my use case is largely the same as yours too (although I can also think of other use cases).

**Comment by dannyhertz** [comment]
Hey! I has similar questions as you did when building my authentication flow in my TCA app. Here is the solution that I settled on and am pretty happy with:

1. Create a networking client that all requests flow through that is stored in your `Environment`. I based my client of the [TinyNetworking](https://github.com/objcio/tiny-networking) micro-library where every request is represented as a simple value type. This approach makes adding new resources super simple and reduces boiler plate a lot. It also works really well with codable models.

2. Now you have a single entry point for all requests and therefore a single place to handle errors, parsing, token juggling, etc. This is the critical point to help you with your current error handling questions. Since every requests goes through this client you now have a single place you can handle auth token application and refreshing (a topic worthy of its own post). 

3. You should now be able to easily expose a publisher to emit all errors emitted by *any* network request. You can then listen for this single publisher at a high level reducer in your app and respond to different types of errors in different ways. Here is how I manage reauth failure errors in my applicationReducer:

```
    Reducer<AppState, AppAction, AppEnvironment> { state, action, environment in
        switch action {
        case let .appDidFinishLaunching(launchOptions):
            return environment.api.reauthorizationFailed()
                .receive(on: environment.mainScheduler)
                .eraseToEffect()
                .map(AppAction.reauthorizationFailed)
                .cancellable(id: AppLifecycleToken())
        case let .reauthorizationFailed(error):
            state = .loggedOut(.init())
            return environment.api
                    .setCredentials(nil)
                    .fireAndForget()
        }
    }
```

In this case `reauthorizationFailed` simply maps the general error publisher and checks for the specific error code represending a reauth failure. I also have another long term subscription in my app reducer that logs ALL errors to our logging system.

Let me know if that is helpful! Happy to share more code!

**Reply by lukeredpath** [reply]
Thanks for sharing - it sounds similar to my idea above but keeping the error publisher encapsulated within the client itself, which sounds interesting. Are you just using a passthrough subject inside your client and forwarding errors through it like in my republishErrors operator above?

**Reply by ericlewis** [reply]
I am curious too, from what it *seems* like, they have an action that creates a long living effect, and then can handle it later.

**Reply by ericlewis** [reply]
actually, I would sort of liken this to more how a location manager works in SWA thinking about it further.

**Reply by lightandshadow68** [reply]
I'm been working on something similar. We have an AppFeature environment that has a SessionClient that publishes changes to authentication state as an enum. When reauthorization fails, it emits .needsAuthentication(LoginCredentials?) that pops authentication in a sheet the user for authentication, pre-populating the user name if provided.

The SessionClient also has a ContentState publisher that indicates when provisioning has completed, the content is "ready" and the User can interact with the UI. These two "tracks" allow the top level view show to orchestrate a "interstitial" screen w/initial login modal, in addition to displaying the application UI w/re-auth modal, etc. Both publishers are long running in that the app subscribes to them when .appDidFinishLaunching(launchOptions): action is received. When the user signs out, the provisioning state is updated, showing the "interstitial" screen again during the de-provisioning process, etc. 

In our use case, provisioning / de-provisioning consists of a number of asynchronous, inter-dependent processes that need to be completed before the content state can switch to .available, so they are handled as Combine pipelines / Effects. An additional publisher emits details about the provisioning process that can be subscribed to and displayed on the "interstitial" screen. 

The SessionClient uses passthrough subjects to send updates to the authentication state, content state and ProvisioningDetails during  each step of pipeline progresses.
